assets/src/models/createDetourMachine.ts

```diff
@@ -1,28 +1,134 @@
-import { setup } from "xstate"
+import { setup, assign } from "xstate"
+import { ShapePoint } from "../schedule"
 
 export const createDetourMachine = setup({
   types: {} as {
-    events: { type: "detour.edit.done" } | { type: "detour.edit.resume" }
+    context: {
+      waypoints: ShapePoint[]
+    }
+    events:
+      | { type: "detour.edit.clear-detour" }
+      | { type: "detour.edit.done" }
+      | { type: "detour.edit.place-connection-point"; location: ShapePoint }
+      | { type: "detour.edit.place-waypoint"; location: ShapePoint }
+      | { type: "detour.edit.resume" }
+      | { type: "detour.edit.undo" }
+      | { type: "detour.share.copy-detour"; detourText: string }
+  },
+  actions: {
+    "detour.add-waypoint": assign({
+      waypoints: ({ context }, params: { location: ShapePoint }) => [
+        ...context.waypoints,
+        params.location,
+      ],
+    }),
+    "detour.remove-last-waypoint": assign({
+      waypoints: ({ context }) => context.waypoints.slice(0, -1),
+    }),
+    "detour.clear": assign({
+      waypoints: [],
+    }),
+
+    "detour.share.copy-detour": (_, { detourText }: { detourText: string }) => {
+      window.navigator.clipboard?.writeText(detourText)
+    },
   },
 }).createMachine({
   id: "Detours Machine",
+  context: () => ({
+    waypoints: [],
+  }),
 
   initial: "Detour Drawing",
   states: {
     "Detour Drawing": {
       initial: "Editing",
       states: {
+        // ? Can this be modeled with a sub state machine as an actor?..
         Editing: {
+          initial: "Pick Start Point",
           on: {
-            "detour.edit.done": {
-              target: "Share Detour",
+            "detour.edit.clear-detour": {
+              target: ".Pick Start Point",
+              actions: "detour.clear",
             },
           },
+          states: {
+            "Pick Start Point": {
+              on: {
+                "detour.edit.place-connection-point": {
+                  target: "Place Waypoint",
+                  actions: {
+                    type: "detour.add-waypoint",
+                    params: ({ event }) => ({
+                      location: event.location,
+                    }),
+                  },
+                },
+              },
+            },
+            "Place Waypoint": {
+              on: {
+                "detour.edit.place-waypoint": {
+                  actions: {
+                    type: "detour.add-waypoint",
+                    params: ({ event }) => ({
+                      location: event.location,
+                    }),
+                  },
+                },
+                "detour.edit.place-connection-point": {
+                  target: "Finished Drawing",
+                  actions: {
+                    type: "detour.add-waypoint",
+                    params: ({ event: { location } }) => ({
+                      location,
+                    }),
```

Comment:
tiny: could the syntax for this function be made identical to the previous references of the action type: detour.add-waypoint? (re-post of [this comment](https://github.com/mbta/skate/pull/2607#discussion_r1610591340))

Reply:
Does this work? https://github.com/mbta/skate/compare/ea2d1939077007fd6f1366a5e775afe13a3b1985..9f4979315bfdd52eeb1fb760a5f985596734b91e

Reply:
totally!